proposals/csharp-8.0/ranges.md

```diff
@@ -133,13 +133,153 @@ var multiDimensional = list[1..2, ..]   // list[Range.Create(1, 2), Range.All]
+
+``` csharp
+List<char> list = ...;
+var value = list[^1]; 
+
+// Gets translated to 
+var value = list[list.Count - 1]; 
+```
+
+The `receiver` and `Length` expressions will be spilled as appropriate to ensure any side effects are only executed once. For example:
+
+``` csharp
+class Collection {
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int this[int index] => _array[index];
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        int i = Get()[^1];
+        Console.WriteLine(i);
+    }
+}
+```
+
+This code will print "Get Length 3". 
+
+### Implicit Range support
+
+The language will provide an instance indexer member with a single parameter of type `Range` for types which meet the following criteria:
+
+- The type is Countable.
+- The type has an accessible member named `Slice` which has two parameters of type `int`.
+- The type does not have an instance indexer which takes a single `Range` as the first parameter. The `Range` must be the only parameter or the remaining parameters must be optional.
+
+For such types, the language will bind as if there is an index member of the form `T this[Range range]` where `T` is the return type of the `Slice` method including any `ref` style annotations. The new member will also have matching accessibility with `Slice`. 
+
+When the `Range` based indexer is bound on an expression named `receiver`, it will be lowered by converting the `Range` expression into two values that are then passed to the `Slice` method. For discussion purposes, lets use the example of `receiver[expr]`.
+
+The first argument of `Slice` will be obtained by converting the typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr2` can be omitted) and `expr1` has type `int`, then it will be emitted as `expr1`.
+- When `expr` is of the form `^expr1..expr2` (where `expr2` can be omitted), then it will be emitted as `receiver.Length - expr1`.
+- When `expr` is of the form `..expr2` (where `expr2` can be omitted), then it will be emitted as `0`.
+- Otherwise, it will be emitted as `expr.Start.GetOffset(receiver.Length)`.
+
+This value will be re-used in the calculation of the second `Slice` argument. When doing so it will be referred to as `start`. The second argument of `Slice` will be obtained by converting the range typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr1` can be omitted) and `expr2` has type `int`, then it will be emitted as `expr2 - start`.
+- When `expr` is of the form `expr1..^expr2` (where `expr1` can be omitted), then it will be emitted as `(receiver.Length - expr2) - start`.
+- When `expr` is of the form `expr1..` (where `expr1` can be omitted), then it will be emitted as `receiver.Length`.
```

Comment:
In all other cases `start` is subtracted to get the number of elements to slice.
```suggestion
- When `expr` is of the form `expr1..` (where `expr1` can be omitted), then it will be emitted as `receiver.Length - start`.
```

Reply:
@pkulikov I think the original is correct. Consider:

```csharp
var slice = someArray[5..];
```

The range would be from `5` to `someArray.Length`, not from `5` to `someArray.Length-5`.

Did I analyze it incorrectly?

Reply:
@BillWagner as far as I understand, the second argument of the `Slice` method is the *number of elements* in the range, not the end of the range. Both the end and start of the range are used to get that number.

In the case of arrays, there is no `Slice` method, but the [RuntimeHelpers.GetSubArray](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.getsubarray?view=netstandard-2.1) method is used. In its [implementation](https://github.com/dotnet/corefx/blob/3c3b9ceee6e1416db6f1ddbc1ab0d0ffb67c16fa/src/Common/src/CoreLib/System/Runtime/CompilerServices/RuntimeHelpers.cs#L19-L51) one of the first lines gets `offset` (which would be `5` in your example) and `length` (which would be `someArray.Length - 5`)

Reply:
Thanks @pkulikov 

After a second read, you are correct. Updated.